lib/commander/runner.rb

```diff
@@ -332,19 +332,27 @@ def remove_global_options(options, args)
       options.each do |option|
         switches = option[:switches]
         next if switches.empty?
-        switch_has_arg = switches.any? { |s| s =~ /[ =]/ }
+        option_has_argument = switches.any? { |s| s =~ /[ =]/ }
         switches = expand_optionally_negative_switches(switches)
 
-        past_switch, arg_removed = false, false
-        args.delete_if do |arg|
-          break if arg == '--'
-          if switches.any? { |s| s[0, arg.length] == arg }
-            arg_removed = !switch_has_arg
-            past_switch = true
-          elsif past_switch && !arg_removed && arg !~ /^-/
-            arg_removed = true
+        option_argument_needs_removal = false
+        args.delete_if do |token|
+          break if token == '--'
+
+          # Use just the portion of the token before the = when
+          # comparing switches.
+          index_of_equals = token.index('=') if option_has_argument
+          token = token[0, index_of_equals] if index_of_equals
+          token_contains_option_argument = !index_of_equals.nil?
+
+          if switches.any? { |s| s[0, token.length] == token }
+            option_argument_needs_removal =
+              option_has_argument && !token_contains_option_argument
+            true
+          elsif option_argument_needs_removal && token !~ /^-/
+            option_argument_needs_removal = false
```

Comment:
Although the logic is definitely cleaner here, I think there is a regression. Now `option_argument_needs_removal` won't be flipped back to false if the global switch is followed by a command switch. For instance, if you add a spec like this to `runner_spec.rb` at line 367 it will pass on master and fail on this branch:

```ruby
it 'should not remove extra values after switches' do
  options = [{ switches: ['--global GLOBAL'] }]
  args = ['--global', '--command', 'arg']
  command_runner.remove_global_options options, args
  expect(args).to eq(['--command', 'arg'])
end
```

In that spec, `args` will be set to `["--command"]` because the code will have removed `"--global"` and `"arg"`.

Possibly all that's needed to fix this is to add `option_argument_needs_removal = false` to the final `else` branch here?

Reply:
Fixed in 2a7b07b